osbuild/objectstore.py

```diff
@@ -64,28 +90,21 @@ def new(self, object_id, base_id=None):
         with tempfile.TemporaryDirectory(dir=self.store) as tmp:
             # the tree that is yielded will be added to the content store
             # on success as object_id
-
-            tree = f"{tmp}/tree"
-            link = f"{tmp}/link"
-            os.mkdir(tree, mode=0o755)
+            tree = TreeObject(f"{tmp}/tree")
 
             if base_id:
                 # the base, the working tree and the output tree are all on
                 # the same fs, so attempt a lightweight copy if the fs
                 # supports it
-                subprocess.run(["cp", "--reflink=auto", "-a", f"{self.refs}/{base_id}/.", tree], check=True)
-
-            yield tree
-
-            # if the yield raises an exception, the working tree is cleaned
-            # up by tempfile, otherwise, we save it in the correct place:
-            fd = os.open(tree, os.O_DIRECTORY)
-            try:
-                m = hashlib.sha256()
-                treesum.treesum(m, fd)
-                treesum_hash = m.hexdigest()
-            finally:
-                os.close(fd)
+                tree.init(f"{self.refs}/{base_id}")
+
+            yield tree.path
+
+            # if `tree.treesum` raises an exception, the working tree
```

Comment:
Treesum can't really raise an exception, no? What we protect against is that `yield does.

Reply:
Ah, I ms-interpreted that comment. I assumed it was talking about possible I/O errors in treesum.

Reply:
That could happen too, you are right. Should apply to both :)